app/src/main/java/com/duckduckgo/widget/SearchAndFavoritesWidget.kt

```diff
@@ -0,0 +1,238 @@
+        extras.putInt(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+        extras.putString(THEME_EXTRAS, widgetTheme.toString())
+
+        val adapterIntent = Intent(context, FavoritesWidgetService::class.java)
+        adapterIntent.putExtras(extras)
+        adapterIntent.data = Uri.parse(adapterIntent.toUri(Intent.URI_INTENT_SCHEME))
+        remoteViews.setRemoteAdapter(R.id.favoritesGrid, adapterIntent)
+        remoteViews.setPendingIntentTemplate(R.id.favoritesGrid, favoriteClickPendingIntent)
+
+        val emptyAdapterIntent = Intent(context, EmptyFavoritesWidgetService::class.java)
+        emptyAdapterIntent.putExtras(extras)
+        emptyAdapterIntent.data = Uri.parse(emptyAdapterIntent.toUri(Intent.URI_INTENT_SCHEME))
+        remoteViews.setEmptyView(R.id.emptyfavoritesGrid, R.id.emptyGridViewContainer)
+        remoteViews.setRemoteAdapter(R.id.emptyfavoritesGrid, emptyAdapterIntent)
+
+        appWidgetManager.updateAppWidget(appWidgetId, remoteViews)
+        appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.favoritesGrid)
+        appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.emptyfavoritesGrid)
+    }
+
+    private fun getLayoutThemed(numColumns: Int, theme: WidgetTheme): Int {
+        return when (theme) {
+            WidgetTheme.LIGHT -> {
+                when (numColumns) {
+                    2 -> R.layout.search_favorites_widget_light_col2
+                    else -> R.layout.search_favorites_widget_light_auto
+                }
+            }
+            WidgetTheme.DARK -> {
+                when (numColumns) {
+                    2 -> R.layout.search_favorites_widget_dark_col2
+                    else -> R.layout.search_favorites_widget_dark_auto
+                }
+            }
+            WidgetTheme.SYSTEM_DEFAULT -> {
+                when (numColumns) {
+                    2 -> R.layout.search_favorites_widget_daynight_col2
+                    else -> R.layout.search_favorites_widget_daynight_auto
+                }
+            }
+        }
+    }
+
+    private fun getCurrentWidgetSize(context: Context, portraitWidth: Int, landsWidth: Int, landsHeight: Int, portraitHeight: Int): Pair<Int, Int> {
+        val orientation = context.resources.configuration.orientation
+        val width = if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
+            landsWidth
+        } else {
+            portraitWidth
+        }
+
+        val height = if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
+            landsHeight
+        } else {
+            portraitHeight
+        }
+
+        var columns = calculateColumns(context, width)
+        var rows = calculateRows(context, height)
+
+        columns = if (columns < 2) 2 else columns
+
+        rows = 1.coerceAtLeast(rows)
+        rows = 4.coerceAtMost(rows)
+
+        return Pair(columns, rows)
+    }
+
+    private fun calculateColumns(context: Context, width: Int): Int {
```

Comment:
We could likely extract `calculateColumns` and `calculateRows` to a utility class which could be unit tested